discretisedfield/field.py

```diff
@@ -3479,3 +3480,210 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        else:
+            geo_units_dict = dict.fromkeys(axes, 'm')
+
+        if self.dim > 1:
+            data_array_dims = axes + ['comp']
+            if self.components is not None:
+                data_array_coords['comp'] = self.components
+            field_array = self.array
+        else:
+            data_array_dims = axes
+            field_array = np.squeeze(self.array, axis=-1)
+
+        data_array = xr.DataArray(field_array,
+                                  dims=data_array_dims,
+                                  coords=data_array_coords,
+                                  name=name,
+                                  attrs=dict(units=units))
+
+        for dim in geo_units_dict:
+            data_array[dim].attrs['units'] = geo_units_dict[dim]
+
+        return data_array
+
+    @classmethod
+    def from_xarray(cls, xa):
+        if not isinstance(xa, xr.DataArray):
+            msg = "Argument must be a xr.DataArray."
+            raise TypeError(msg)
+
+        if xa.ndim not in [3, 4]:
+            msg = "DataArray dimensions must be 3 or 4."
+            raise ValueError(msg)
+
+        if xa.ndim == 3:
+            for dim in xa.dims:
+                if dim not in ['x', 'y', 'z']:
+                    msg = "The dimensions must be either 'x', 'y', or 'z'."
+                    raise ValueError(msg)
+        elif xa.ndim == 4:
+            for dim in xa.dims:
+                if dim not in ['x', 'y', 'z', 'comp']:
+                    msg = "The dimensions must be either "\
+                          "'x', 'y', 'z' or 'comp'."
+                    raise ValueError(msg)
+            if xa['comp'].values.size != len(set(xa['comp'].values)):
+                msg = "All the components should be unique."
+                raise ValueError(msg)
+            if 'comp' in xa.coords:
+                for comp in xa['comp'].values:
+                    if not isinstance(comp, str):
+                        msg = "The values of 'comp' must be a string."
+                        raise ValueError(msg)
+                    elif comp not in ['x', 'y', 'z']:
+                        msg = "The values of 'comp' must be either "\
+                              "'x', 'y', or 'z'."
+                        raise ValueError(msg)
+
+        for i in 'xyz':
+            if xa[i].values.size != 1:
+                if not np.allclose(np.diff(xa[i].values),
+                                   np.diff(xa[i].values).mean()):
+                    msg = f'Co-ordinates of {i} must be'\
+                          ' equally spaced.'
+                    raise ValueError(msg)
+
+        if xa.values.dtype not in [np.float64, np.complex128]:
+            msg = "DataArray values must be either "\
+                  "np.float64 or np.complex128."
+            raise ValueError(msg)
+
+        cell = tuple()
+        p1 = tuple()
+        p2 = tuple()
+        for i in 'xyz':
+            dec = 5  # minimum decimal places to round to.
```

Comment:
Does this work for nanometers?

Reply:
Yes, it works for nanometers because of:
```Python
exp = np.log10(np.abs(diff))
if exp < 0:
    dec = dec + int(np.abs(exp).round())
```
Here, the code calculates the exponent of the `diff` (essentially the cell dimension) and adds it to `dec`. So, for nanometer dimensions, `dec = 5 + 9` which is 14.

Reply:
Using `dec = np.finfo(xa[i].values.dtype).precision` now as suggested by @lang-m.